Add flatten for named tensors.

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25658 Implement named tensor `unflatten(dim, namedshape)`.
* **#25672 Add flatten for named tensors.**

There are three overloads:
1) flatten(tensor, int start_dim, int end_dim, Dimname out_dim)
2) flatten(tensor, Dimname start_dim, Dimname end_dim, Dimname out_dim)
3) flatten(tensor, DimnameList dims, Dimname out_dim)

`flatten` joins all the dimensions between start_dim and end_dim into
one dimension. The name of the output dimension is specified by
`out_dim`.

In the case where flatten takes a list of `dims` to flatten, all the
dimensions in `dims` must be in consecutive order.

Test Plan:
- new tests [namedtensor ci]

Differential Revision: [D17192656](https://our.internmc.facebook.com/intern/diff/D17192656)

zou3519 committed:
Add flatten for named tensors.

There are three overloads:
1) flatten(tensor, int start_dim, int end_dim, Dimname out_dim)
2) flatten(tensor, Dimname start_dim, Dimname end_dim, Dimname out_dim)
3) flatten(tensor, DimnameList dims, Dimname out_dim)

`flatten` joins all the dimensions between start_dim and end_dim into
one dimension. The name of the output dimension is specified by
`out_dim`.

In the case where flatten takes a list of `dims` to flatten, all the
dimensions in `dims` must be in consecutive order.

Test Plan:
- new tests [namedtensor ci]

zou3519 committed:
Update on "Add flatten for named tensors."


There are three overloads:
1) flatten(tensor, int start_dim, int end_dim, Dimname out_dim)
2) flatten(tensor, Dimname start_dim, Dimname end_dim, Dimname out_dim)
3) flatten(tensor, DimnameList dims, Dimname out_dim)

`flatten` joins all the dimensions between start_dim and end_dim into
one dimension. The name of the output dimension is specified by
`out_dim`.

In the case where flatten takes a list of `dims` to flatten, all the
dimensions in `dims` must be in consecutive order.

Test Plan:
- new tests [namedtensor ci]

nairbv commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -942,6 +942,19 @@
   variants: function, method
+  named_guard: False
+
+- func: flatten(Tensor self, int start_dim, int end_dim, Dimname out_dim) -> Tensor
```
Would it make sense for `out_dim`s to be optional, defaulting to wildcard?

Would a mixed interface also be useful, something like `Dimname start_dim, int end_dim=-1`

zou3519 replied:
> Would it make sense for out_dims to be optional, defaulting to wildcard?

I think this is too much magic and the API doesn't read well. For example, for Tensor[N, C, H, W], `tensor.flatten('C', out_dim='features')` doesn't give the reader too much useful information.

> Would a mixed interface also be useful

Mixed interface is annoying to support. It's useful if the user has unnamed dimensions and have trouble referring to them, but then the workaround is to just use the `flatten(int, int, Dimname)` overload. We don't support mixed anywhere else (for example, torch.sum(...) takes a DimnameList or an IntList, but not mixed.)

nairbv replied:
partly I ask because `flatten(Tensor self, int start_dim=0, int end_dim=-1)` exists, so seems natural to have a named equivalent. Is it necessary to always require the out_dim to be named or could it default to wildcard?

zou3519 replied:
I'm going to require the `out_dim` to be named for now. If someone is using named tensors, it's better to steer them to having more named dimensions than fewer.

zou3519 committed:
Update on "Add flatten for named tensors."


There are three overloads:
1) flatten(tensor, int start_dim, int end_dim, Dimname out_dim)
2) flatten(tensor, Dimname start_dim, Dimname end_dim, Dimname out_dim)
3) flatten(tensor, DimnameList dims, Dimname out_dim)

`flatten` joins all the dimensions between start_dim and end_dim into
one dimension. The name of the output dimension is specified by
`out_dim`.

In the case where flatten takes a list of `dims` to flatten, all the
dimensions in `dims` must be in consecutive order.

Test Plan:
- new tests [namedtensor ci]

Differential Revision: [D17192656](https://our.internmc.facebook.com/intern/diff/D17192656)
